src/metabase/pulse/render/body.clj

```diff
@@ -512,14 +512,51 @@
       [:img {:style (style/style {:display :block :width :100%})
              :src   (:image-src image-bundle)}]]}))
 
+(defn- largest-range
+  [{:keys [fingerprint]}]
+  (let [{mn :min mx :max} (get-in fingerprint [:type :type/Number])]
+    (- mx mn)))
+
+(defn- overlap
+  [col-a col-b]
+  (let [[col-a col-b] (reverse (sort-by largest-range [col-a col-b]))
+        [min-a min-b] (map #(get-in % [:fingerprint :type :type/Number :min]) [col-a col-b])
+        [max-a max-b] (map #(get-in % [:fingerprint :type :type/Number :max]) [col-a col-b])
+        [a b c d] (sort [min-a min-b max-a max-b])
+        max-width (- d a)
+        overlap-width (- c b)]
+    (cond
+      ;; the smaller range is completely ahead of the larger range
+      (and (< max-a min-b) (< max-a max-b)) 0
+      ;; the smaller range is completely behind the larger range
+      (and (> min-a min-b) (> min-a max-b)) 0
+      ;; there is some overlap, so what's the percentage overlap of the whole width of both ranges?
+      :else (/ overlap-width max-width))))
+
+(defn- group-axes
+  [cols-meta group-threshold]
+  (let [cols-by-type (group-by (juxt :base_type :effective_type :semantic_type) cols-meta)]
+    (when (not= (count cols-by-type) (count cols-meta))
+      (let [num? (fn [[k _]] (isa? (first k) :type/Number))
```

Comment:
soft preference for either `[k]` or `[k _v]`. Or really `[[[col-type]]] (isa? col-type :type/Number)`.